File: Assignment/Chapter 1/Student Grade Calculator/Dart Version/lib/core/services/export/csv_report_exporter.dart
import '../../models/export_artifact.dart';
import '../../models/export_format.dart';
import '../../models/processing_report.dart';
import 'base_report_exporter.dart';

class CsvReportExporter extends BaseReportExporter {
  const CsvReportExporter();

  @override
  ExportFormat get format => ExportFormat.csv;

  @override
  Future<ExportArtifact> export(
    ProcessingReport report,
    String destinationPath,
  ) {
    final rows = [gradeHeaders, ...buildGradeRows(report)];
    final csv = rows.map(_encodeRow).join('\n');
    return writeText(csv, destinationPath);
  }

  String _encodeRow(List<String> row) => row
      .map((value) => '"${value.replaceAll('"', '""')}"')
      .join(',');
}

File: Assignment/Chapter 1/Student Grade Calculator/Kotlin Version/app/src/main/java/com/romualdsigning/studentgradecalc/data/export/CsvReportExporter.kt
package com.romualdsigning.studentgradecalc.data.export

import android.content.Context
import android.net.Uri
import com.github.doyaaaaaken.kotlincsv.dsl.csvWriter
import com.romualdsigning.studentgradecalc.domain.model.ExportArtifact
import com.romualdsigning.studentgradecalc.domain.model.ExportDestination
import com.romualdsigning.studentgradecalc.domain.model.ExportFormat
import com.romualdsigning.studentgradecalc.domain.model.ProcessingReport

class CsvReportExporter : AbstractReportExporter() {
    override val format: ExportFormat = ExportFormat.CSV

    override suspend fun export(
        context: Context,
        report: ProcessingReport,
        destination: Uri,
        exportDestination: ExportDestination,
    ): ExportArtifact {
        val csv = csvWriter().writeAllAsString(listOf(gradeHeaders) + buildGradeRows(report))
        return writeBytes(context, destination, csv.toByteArray(), exportDestination)
    }
}
